jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt

```diff
@@ -6,64 +6,121 @@ package software.aws.toolkits.jetbrains.core
+     */
+    fun <T> getResource(
+        resource: CachedResource<T>,
+        region: AwsRegion,
+        credentialProvider: ToolkitCredentialsProvider,
+        useStale: Boolean = true,
+        forceFetch: Boolean = false
+    ): CompletionStage<T>
+
+    /**
+     * Clears the contents of the cache across all regions, credentials and resource types.
+     */
+    fun clear()
 
     companion object {
-        fun getInstance(project: Project): AwsResourceCache =
-            ServiceManager.getService(project, AwsResourceCache::class.java)
+        @JvmStatic
+        fun getInstance(project: Project): AwsResourceCache = ServiceManager.getService(project, AwsResourceCache::class.java)
     }
 }
 
-class DefaultAwsResourceCache(
-    private val accountSettingsManager: ProjectAccountSettingsManager,
-    private val clientManager: ToolkitClientManager
-) : AwsResourceCache {
-    private val cache = ConcurrentHashMap<String, Any>()
-
-    @Suppress("UNCHECKED_CAST")
-    override fun lambdaFunctions(): CompletableFuture<List<LambdaFunction>> {
-        val credentialProvider = try {
-            accountSettingsManager.activeCredentialProvider
-        } catch (_: Exception) {
-            return CompletableFuture.completedFuture(emptyList())
-        }
-
-        val region = accountSettingsManager.activeRegion
-        val credentialProviderId = credentialProvider.id
+interface CachedResource<T> {
+    fun fetch(project: Project, region: AwsRegion, credentials: ToolkitCredentialsProvider): T
+}
 
-        val resourceKey = "$region:$credentialProviderId:lambdafunctions"
+class DefaultAwsResourceCache(private val project: Project, private val clock: Clock) : AwsResourceCache {
+    @Suppress("unused")
+    constructor(project: Project) : this(project, Clock.systemDefaultZone())
 
-        val cachedResult = cache[resourceKey]
-        cachedResult?.let {
-            return CompletableFuture.completedFuture(cachedResult as List<LambdaFunction>)
-        }
+    private val cache = ConcurrentHashMap<Key<*>, Entry<*>>()
+    private val accountSettings = ProjectAccountSettingsManager.getInstance(project)
 
-        val resultFuture = CompletableFuture<List<LambdaFunction>>()
-        val client = clientManager.getClient<LambdaClient>()
+    override fun <T> getResource(resource: CachedResource<T>, useStale: Boolean, forceFetch: Boolean) =
+        getResource(resource, accountSettings.activeRegion, accountSettings.activeCredentialProvider, useStale, forceFetch)
 
+    override fun <T> getResource(
+        resource: CachedResource<T>,
+        region: AwsRegion,
+        credentialProvider: ToolkitCredentialsProvider,
+        useStale: Boolean,
+        forceFetch: Boolean
+    ): CompletionStage<T> {
+        val key = Key(resource, region, credentialProvider)
+        val future = CompletableFuture<T>()
         ApplicationManager.getApplication().executeOnPooledThread {
```

Comment:
Should we check for cache hit before off loading the thread? This way its faster since no need to switch threads

Reply:
Opted for a simpler implementation for now both on the part of the caller (no behavioural difference between a cache hit and cache miss - just lower latency) and the cache itself (easier handling of the cache map, fewer locking concerns). Whilst I agree that it's potentially more efficient to stay in the same thread to check the map - at this point I don't think the optimization is worth the complexity.

Reply:
Finneee